internal/logger/logger.go

```diff
@@ -0,0 +1,71 @@
+/**
+ * Copyright (c) F5, Inc.
+ *
+ * This source code is licensed under the Apache License, Version 2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+package logger
+
+import (
+	"io"
+	"log/slog"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/nginx/agent/v3/internal/config"
+)
+
+const (
+	defaultLogFile = "agent.log"
+)
+
+var logLevels = map[string]slog.Level{
+	"debug": slog.LevelDebug,
+	"info":  slog.LevelInfo,
+	"warn":  slog.LevelWarn,
+	"error": slog.LevelError,
+}
+
+func New(params config.Log) *slog.Logger {
+	handler := slog.NewTextHandler(
+		getLogWriter(params.Path),
+		&slog.HandlerOptions{
+			Level: getLogLevel(params.Level),
+		},
+	)
+
+	return slog.New(handler)
+}
+
+func getLogLevel(level string) slog.Level {
+	if level == "" {
+		return slog.LevelInfo
+	}
+	return logLevels[strings.ToLower(level)]
+}
+
+func getLogWriter(logFile string) io.Writer {
+	logPath := logFile
+	if logFile != "" {
+		fileInfo, err := os.Stat(logPath)
+		if err != nil {
+			slog.Error("error reading log directory, proceeding to log only to stdout/stderr", "error", err)
+			return os.Stderr
+		}
+
+		if fileInfo.IsDir() {
+			logPath = path.Join(logPath, defaultLogFile)
+		}
+
+		logFileHandle, err := os.OpenFile(logPath, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0o600)
+		if err != nil {
+			slog.Error("failed to open log file, proceeding to log only to stdout/stderr", "error", err)
+			return os.Stderr
+		}
+		return logFileHandle
```

Comment:
Since the `logFile` argument comes from user input, this code will likely return an error from our static analyzer for Path Traversal Attacks: https://owasp.org/www-community/attacks/Path_Traversal

We'd need to resolve symlinks and ensure that we are working in a trusted path to avoid that error. Here's a good Go guide for avoiding it: https://www.stackhawk.com/blog/golang-path-traversal-guide-examples-and-prevention/

Reply:
For log files I think allowing the user to freely specify where they want their logs to be written to is ok, but we will be adding an `allowedDirectories` map in another PR that we can use to validate config file paths for NGINX, Unit, etc.